EIPS/eip-8037.md

```diff
@@ -103,6 +174,76 @@ This proposal is consistent with the multidimensional gas metering introduced in
 - What counts as “same code”? Exact runtime bytecode. Even minor differences produce distinct hashes.
 - Empty code handling: Clients can treat empty code as a special case with a hard-coded hash lookup (`EMPTY_CODE_HASH`), making it effectively free.
 
+#### Access-List CodeHash Deduplication Rationale (Solving the existing-bytecode lookup issues)
+
+The access-list based deduplication mechanism addresses a critical consensus issue that arises from database divergence across different node implementations and sync modes:
+
+##### The Database Divergence Problem
+
+Empirical analysis reveals that approximately **27,869 bytecodes** exist in full-synced node databases with no live account pointing to them (as of the time of writing this EIP). These orphaned codes arise from:
+
+1. Chain reorganizations: Code blobs persist in the database even when the blocks containing their creation are reorganized away
+2. Sync mode differences:
+   - Full-sync nodes: Retain all historical code, including from reverted/reorged transactions
+   - Snap-sync nodes: Only store code reachable from the current state trie
+   - Archive nodes: May have different retention policies
+
+This divergence means that a naive "database contains hash?" check would yield **different gas costs on different nodes for the same transaction**, breaking consensus.
+
+##### Why Access-List Based Solution
+
+The access-list approach solves the bytecode duplication check problem by:
+
+1. Deterministic behavior: Only code hashes explicitly included in the transaction's access list are considered for deduplication. This makes the result independent of local database state.
+2. No reverse index requirement: Unlike other approaches, this doesn't require maintaining a `codeHash → account` reverse index, which would add significant complexity and storage overhead.
+3. Leverages existing infrastructure: Builds on EIP-2930 access lists and EIP-2929 access costs, requiring minimal protocol changes.
+4. Explicit opt-in: Transactions must explicitly indicate which addresses to check, preventing unexpected behavior and giving users control over gas optimization.
+5. Forward compatibility: Older clients that don't implement the feature simply never grant the discount, maintaining compatibility while newer clients can optimize. Same will happen with Wallet providers and TX builders.
+
+#### Access-List Extension Specification
```

Comment:
This section shouldn't be under "rationale". We should create a section named "Reference Implementation" . Check  [EIP-1](https://eips.ethereum.org/EIPS/eip-1#what-belongs-in-a-successful-eip)